### PR TITLE
Communicate with bugfinder backend

### PR DESCRIPTION
**Pull request by IlijazM**
Closes Gamify-IT/issues#290

### CODE REVIEW

File: docker-compose-dev.yaml [hunk 1]
```diff
@@ -1,124 +1,115 @@
...
-  #      - POSTGRES_PASSWORD=postgres
+  login-db:
+    container_name: login-db
+    image: postgres:14-alpine
+    restart: always
+    expose:
+      - '5432'
+    environment:
+      - POSTGRES_USER=postgres
+      - POSTGRES_PASSWORD=postgres
```
Comment (LeviUniAcc) [inline docker-compose-dev.yaml:13]:
```suggestion
```

File: docker-compose-dev.yaml [hunk 2]
```diff
@@ -1,124 +1,115 @@
...
+    container_name: login-backend
+    image: ghcr.io/gamify-it/login-backend:main
+    restart: always
+    pull_policy: always
+    expose:
+      - '80'
+    environment:
+      - JWT_KEY=SecretKEY
+      - JWT_VALIDITY_DURATION=720h
+      - POSTGRES_URL=postgresql://postgres:postgres@login-db:5432/
```
Comment (LeviUniAcc) [inline docker-compose-dev.yaml:46]:
```suggestion
```

File: docker-compose-dev.yaml [hunk 3]
```diff
@@ -1,124 +1,115 @@
...
-  #   pull_policy: always
-  #   expose:
-  #     - "80"
+  login-frontend:
+    container_name: login-frontend
+    image: ghcr.io/gamify-it/login-frontend:main
+    restart: always
+    pull_policy: always
+    expose:
+      - '80'
```
Comment (LeviUniAcc) [inline docker-compose-dev.yaml:69]:
```suggestion
```

File: docker-compose.yaml [hunk 4]
```diff
@@ -1,129 +1,117 @@
...
-  #     - POSTGRES_PASSWORD=postgres
+  login-db:
+    container_name: login-db
+    image: postgres:14-alpine
+    restart: always
+    expose:
+      - '5432'
+    environment:
+      - POSTGRES_USER=postgres
+      - POSTGRES_PASSWORD=postgres
```
Comment (LeviUniAcc) [inline docker-compose.yaml:13]:
```suggestion
```

File: docker-compose.yaml [hunk 5]
```diff
@@ -1,129 +1,117 @@
...
+    container_name: login-backend
+    image: ghcr.io/gamify-it/login-backend:main
+    restart: always
+    pull_policy: always
+    expose:
+      - '80'
+    environment:
+      - JWT_KEY=SecretKEY
+      - JWT_VALIDITY_DURATION=720h
+      - POSTGRES_URL=postgresql://postgres:postgres@login-db:5432/
```
Comment (LeviUniAcc) [inline docker-compose.yaml:46]:
```suggestion
```

File: docker-compose.yaml [hunk 6]
```diff
@@ -1,129 +1,117 @@
...
-  #   pull_policy: always
-  #   expose:
-  #     - "80"
+  login-frontend:
+    container_name: login-frontend
+    image: ghcr.io/gamify-it/login-frontend:main
+    restart: always
+    pull_policy: always
+    expose:
+      - '80'
```
Comment (LeviUniAcc) [inline docker-compose.yaml:71]:
```suggestion
```

File: docker-compose.yaml [hunk 7]
```diff
@@ -1,129 +1,117 @@
...
+    depends_on:
+      - overworld-db
+    environment:
+      - POSTGRES_USER=postgres
+      - POSTGRES_PASSWORD=postgres
+      - POSTGRES_URL=postgresql://overworld-db:5432/postgres
 
+  ## TODO: insert bugfinder backend when ready
```
Comment (LeviUniAcc) [inline docker-compose.yaml:62]: cant test until bugfinder backend is inserted... (pls do so in all relevante compose files

**Review by LeviUniAcc — COMMENTED** [review]
i cant find a single REST request? i dont get this pull request at all, not a single change fits to the pr name

File: src/services/bugfindergame.ts [hunk 8]
```diff
@@ -114,4 +122,18 @@ export class BugFinderGame {
...
+   */
+  private async fetchCurrentCode() {
+    const configuration = window.location.pathname.split('/').pop();
+    if (configuration == null) {
+      throw Error('No configuration selected!');
+    }
+    const res = await fetch(`${BASE_URL}/configuration/${configuration}/codes`);
+    const json = (await res.json()) as ICode[];
+    this.currentCode = json[this.currentCodeNumber];
+    return this.currentCode;
```
Comment (LeviUniAcc) [inline src/services/bugfindergame.ts:131]: pls adjust REST requests like finitequiz/ lecturer interface for consistancy

File: src/main.ts [hunk 9]
```diff
@@ -8,9 +8,42 @@ import 'bootstrap-vue-3/dist/bootstrap-vue-3.css';
...
+    })
+  );
+  await Promise.all(
+    solutionJson.map(async (solution) => {
+      await fetch(`${BASE_URL}/solution`, {
+        method: 'POST',
+        body: JSON.stringify(solution),
+        headers,
+      });
+    })
```
Comment (LeviUniAcc) [inline src/main.ts:71]: pls adjust REST requests like finitequiz/ lecturer interface for consistancy

File: coverage/clover.xml [hunk 10]
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (delvh) [inline coverage/clover.xml:1]: Remove this directory from the repo

**Comment by LeviUniAcc** [comment]
grafik
cant test while test data image not found

**Comment by maexled** [comment]
please re-review

**Comment by maexled** [comment]
How to test:

Edit docker-compose-dev.yaml:
- set setup.image to `ghcr.io/gamify-it/test-data:maintenance-remove-duplicate-configuration`
- set bugfinder-backend.image to `ghcr.io/gamify-it/bugfinder-backend:feature-bugfinder-backend`

Run `docker-compose -f docker-compose-dev.yaml up`
Start bugfinder frontend: `npm run serve`
Open `http://localhost` and log in
Go to minigame spot 4 and play bugfinder: There will be 5 code snippets loaded by the backend
grafik

**Comment by cdingm** [comment]
> How to test:
> 
> Edit docker-compose-dev.yaml:
> 
>     * set setup.image to `ghcr.io/gamify-it/test-data:maintenance-remove-duplicate-configuration`
> 
>     * set bugfinder-backend.image to `ghcr.io/gamify-it/bugfinder-backend:feature-bugfinder-backend`
> 
> 
> Run `docker-compose -f docker-compose-dev.yaml up` Start bugfinder frontend: `npm run serve` Open `http://localhost` and log in Play course PSE Go to minigame spot 4 and play bugfinder: There will be 5 code snippets loaded by the backend grafik

this worked for me, results got posted (201)

File: src/models/code.ts [hunk 11]
```diff
@@ -29,47 +29,47 @@ export enum ErrorType {
...
-  public constructor(id: number, words: IWord[]) {
+  public constructor(id: string, words: IWord[]) {
     this.id = id;
     this.words = words;
   }
 }
 
 export class Solution implements ISolution {
-  id: number;
+  id?: string;
```
Comment (cdingm) [inline src/models/code.ts:58]: is the "?" a typo?

**Review by LeviUniAcc — APPROVED** [review]
tryed it like described and it worked

File: docker-compose-dev-e2e.yaml [hunk 12]
```diff
@@ -1,100 +1,108 @@
...
-  #   pull_policy: always
-  #   restart: on-failure
-  #   environment:
-  #     POSTGRES_SETUP: >
-  #       login/default.sql
-  #       overworld/default.sql
-  #       bugfinder/default.sql
+  setup:
+    container_name: setup
+    image: ghcr.io/gamify-it/test-data:feature-bugfinder-dump
```
Comment (Aaron2000119) [inline docker-compose-dev-e2e.yaml:7]:
```suggestion
    image: ghcr.io/gamify-it/test-data:latest
```

File: docker-compose-dev.yaml [hunk 13]
```diff
@@ -1,124 +1,134 @@
...
-  #    pull_policy: always
-  #    restart: on-failure
-  #    environment:
-  #      POSTGRES_SETUP: >
-  #        login/default.sql
-  #        overworld/default.sql
-  #        bugfinder/default.sql
+  setup:
+    container_name: setup
+    image: ghcr.io/gamify-it/test-data:feature-bugfinder-dump
```
Comment (Aaron2000119) [inline docker-compose-dev.yaml:7]:
```suggestion
    image: ghcr.io/gamify-it/test-data:latest
```

File: docker-compose.yaml [hunk 14]
```diff
@@ -1,129 +1,142 @@
...
-  #   pull_policy: always
-  #   restart: on-failure
-  #   environment:
-  #     POSTGRES_SETUP: >
-  #       login/default.sql
-  #       overworld/default.sql
-  #       bugfinder/default.sql
+  setup:
+    container_name: setup
+    image: ghcr.io/gamify-it/test-data:feature-bugfinder-dump
```
Comment (Aaron2000119) [inline docker-compose.yaml:7]:
```suggestion
    image: ghcr.io/gamify-it/test-data:latest
```

**Review by Gr33ndev — APPROVED** [review]

File: src/services/bugfindergame.ts [hunk 15]
```diff
@@ -112,6 +125,54 @@ export class BugFinderGame {
...
+    return this.codeCache;
+  }
+
+  /**
+   * fetches the solution for the current code
+   */
+  private async fetchSolution() {
+    const res = await fetch(`${BASE_URL}/configurations/${this.configuration}/codes/${this.currentCode?.id}/solutions`);
+    const json = (await res.json()) as ISolution;
+    return json;
```
Comment (Aaron2000119) [inline src/services/bugfindergame.ts:176]:
```suggestion
    return (await res.json()) as ISolution;
```
can be simplified

**Review by Aaron2000119 — CHANGES_REQUESTED** [review]

**Pull request merged**